src/main/java/com/epam/test_generator/services/JiraService.java

```diff
@@ -0,0 +1,343 @@
+            "project in (%s) AND status in (Open, \"In Progress\", Reopened, Verified) AND type=sub-task",
+            projects);
+
+        List<JiraStory> jiraStories = jiraStoryDAO.getJiraStoriesByFilter(query);
+        List<JiraSubTask> jiraSubTasks = jiraSubStoryDAO.getJiraSubtoriesByFilter(subtaskQuery);
+
+        Map<String, Suit> suits = suitDAO.findAll().stream().filter(suit ->
+            jiraStories.stream().filter(jiraStory -> jiraStory.getJiraKey() != null).noneMatch(jiraStory -> jiraStory.getJiraKey().equals(suit.getJiraKey())))
+            .collect(Collectors.toMap(Suit::getJiraKey, suit -> suit));
+
+        Map<String, Case> cases = caseDAO.findAll().stream().filter(caze ->
+            jiraSubTasks.stream().filter(jiraSubTask -> jiraSubTask.getJiraKey() != null).noneMatch(jiraSubTask -> jiraSubTask.getJiraKey().equals(caze.getJiraKey())))
+            .collect(Collectors.toMap(Case::getJiraKey, caze -> caze));
+
+
+
+        jiraStories.forEach(this::updateSuitFromJiraStory);
+
+        jiraSubTasks.stream()
+            .filter(subTask -> suitDAO.findByJiraKey(subTask.getJiraParentKey()) != null)
+            .forEach(subTask -> {
+                Case aCase = caseDAO.findByJiraKey(subTask.getJiraKey());
+                if (aCase == null) {
+                    createCaseFromJiraSubTask(subTask);
+                } else {
+                    updateSubTaskFromJiraSubTask(subTask);
+                }
+            });
+
+        suits.forEach((jiraKey, suit) -> {
+            projectDAO.findByJiraKey(suit.getJiraProjectKey()).getSuits().remove(suit);
+            suitDAO.delete(suit);
+        });
+        cases.forEach((jiraKey, caze) -> {
+            suitDAO.findByJiraKey(caze.getJiraParentKey()).getCases().remove(caze);
+            caseDAO.delete(caze);
+        });
+
+    }
+
+
+    /**
+     * Synchronisation method to Jira The method puts all changes in BBD (all changed suits=stories,
+     * cases=sub stories) to Jira in case when suit or case has deleted - stories/sub stories will
+     * be closed in Jira too
+     */
+    public void syncToJira() throws JiraException {
+
+        for (Suit suit : suitDAO.findAll()) {
+            if (suit.getLastJiraSyncDate() == null) {
+                jiraStoryDAO.createStory(suit);
+                for (Case cases : suit.getCases()) {
+                    jiraSubStoryDAO.createSubStory(cases);
+                }
+
+            } else if (suit.getLastModifiedDate().isAfter(suit.getLastJiraSyncDate())) {
+                jiraStoryDAO.updateStoryByJiraKey(suit);
+            }
+            for (Case caze : suit.getCases()) {
+                if (caze.getLastJiraSyncDate() == null) {
+                    jiraSubStoryDAO.createSubStory(caze);
+                } else if (caze.getLastModifiedDate().isAfter(caze.getLastJiraSyncDate())) {
+                    jiraSubStoryDAO.updateSubStoryByJiraKey(caze);
+                }
+            }
+        }
```

Comment:
Another developer will spend some time trying to understand the logic here.
It is not a good idea to make high level of nesting of loops and if-blocks.

Read Fowler's article about extracting method [here](https://refactoring.com/catalog/extractMethod.html) or in his book Refactoring.

Kindly try to extract some sub steps to private methods and give them clear names.